src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+        /// </summary>
+        protected TLocalState State;
+        protected TLocalState StateWhenTrue;
+        protected TLocalState StateWhenFalse;
+        protected bool IsConditionalState;
+
+        private readonly bool _nonMonotonicTransfer; // for the data flows out walker, we track unassignments as well as assignments
+
+        protected void SetConditionalState(TLocalState whenTrue, TLocalState whenFalse)
         {
-            this._trackUnassignments = trackUnassignments;
+            IsConditionalState = true;
+            State = default(TLocalState);
+            StateWhenTrue = whenTrue;
+            StateWhenFalse = whenFalse;
         }
 
+        protected void SetState(TLocalState newState)
+        {
+            StateWhenTrue = StateWhenFalse = default(TLocalState);
+            IsConditionalState = false;
+            State = newState;
+        }
+
+        protected void Split()
+        {
+            Debug.Assert(!_trackExceptions || PendingBranches[0].Branch == null);
+            if (!IsConditionalState)
+            {
+                SetConditionalState(State, State.Clone());
+            }
+        }
+
+        protected void Unsplit()
+        {
+            Debug.Assert(!_trackExceptions || PendingBranches[0].Branch == null);
+            if (IsConditionalState)
+            {
+                Join(ref StateWhenTrue, ref StateWhenFalse);
+                SetState(StateWhenTrue);
+            }
+        }
+
+        /// <summary>
+        /// Where all diagnostics are deposited.
+        /// </summary>
+        protected DiagnosticBag Diagnostics { get; }
+
+        #region Region
+        // For region analysis, we maintain some extra data.
+        protected enum RegionPlace { Before, Inside, After };
+        protected RegionPlace regionPlace; // tells whether we are currently analyzing code before, during, or after the region
+        protected readonly BoundNode firstInRegion, lastInRegion;
+        private readonly bool _trackRegions;
+
+        /// <summary>
+        /// A cache of the state at the backward branch point of each loop.  This is not needed
+        /// during normal flow analysis, but is needed for DataFlowsOut region analysis.
+        /// </summary>
+        private readonly Dictionary<BoundLoopStatement, TLocalState> _loopHeadState;
+        #endregion Region
+
         protected AbstractFlowPass(
             CSharpCompilation compilation,
             Symbol member,
             BoundNode node,
-            BoundNode firstInRegion,
-            BoundNode lastInRegion,
-            bool trackRegions = true,
-            bool trackUnassignments = false)
-            : base(compilation, member, node, firstInRegion, lastInRegion, trackRegions)
+            BoundNode firstInRegion = null,
+            BoundNode lastInRegion = null,
+            bool trackRegions = false,
+            bool trackExceptions = false,
+            bool nonMonotonicTransferFunction = false)
```

Comment:
do we expect subclasses to make use of all these optional parameters (i.e. each subclass using a different subset of parameters)?  Or is it more like some subclasses will just pass the required params, and some will pass all the optional params.  If the latter, this might be cleaner as separate constructors with required params that all defer to a single private construct with nothing being optional.